File: erlymon-monitor-app/src/main/java/org/erlymon/monitor/ui/activity/MainActivity.java
/*
 * Copyright (c) 2016, Sergey Penkovsky <sergey.penkovsky@gmail.com>
 *
 * This file is part of Erlymon Monitor.
 *
 * Erlymon Monitor is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Erlymon Monitor is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Erlymon Monitor.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.erlymon.monitor.ui.activity;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.arellomobile.mvp.MvpAppCompatActivity;
import com.arellomobile.mvp.presenter.InjectPresenter;
import com.squareup.otto.Bus;

import org.erlymon.monitor.MainApp;
import org.erlymon.monitor.R;
import org.erlymon.monitor.Screens;
import org.erlymon.monitor.mvp.model.Server;
import org.erlymon.monitor.mvp.model.User;
import org.erlymon.monitor.mvp.presenter.CheckSessionPresenter;
import org.erlymon.monitor.mvp.presenter.GetServerPresenter;
import org.erlymon.monitor.mvp.presenter.SignOutPresenter;
import org.erlymon.monitor.mvp.view.CheckSessionView;
import org.erlymon.monitor.mvp.view.GetServerView;
import org.erlymon.monitor.mvp.view.SignOutView;
import org.erlymon.monitor.ui.fragment.AboutFragment;
import org.erlymon.monitor.ui.fragment.DeviceFragment;
import org.erlymon.monitor.ui.fragment.DevicesFragment;
import org.erlymon.monitor.ui.fragment.MapFragment;
import org.erlymon.monitor.ui.fragment.PermissionsFragment;
import org.erlymon.monitor.ui.fragment.ReportFragment;
import org.erlymon.monitor.ui.fragment.ServerFragment;
import org.erlymon.monitor.ui.fragment.SignInFragment;
import org.erlymon.monitor.ui.fragment.SignUpFragment;
import org.erlymon.monitor.ui.fragment.UserFragment;
import org.erlymon.monitor.ui.fragment.UsersFragment;
import org.osmdroid.tileprovider.tilesource.bing.BingMapTileSource;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import ru.terrakok.cicerone.Navigator;
import ru.terrakok.cicerone.NavigatorHolder;
import ru.terrakok.cicerone.Router;
import ru.terrakok.cicerone.android.SupportFragmentNavigator;
import timber.log.Timber;

public class MainActivity extends MvpAppCompatActivity
        implements NavigationView.OnNavigationItemSelectedListener,
        SignInFragment.OnViewListener,
        GetServerView,
        CheckSessionView,
        SignOutView {
    @BindView(R.id.toolbar)
    Toolbar mToolbar;

    @BindView(R.id.progress)
    ProgressBar mProgress;

    @BindView(R.id.container)
    FrameLayout mContainer;

    @BindView(R.id.drawer_layout)
    DrawerLayout drawer;

    @BindView(R.id.nav_view)
    NavigationView navigationView;

    TextView mLabelName;

    TextView mLabelEmail;

    @InjectPresenter
    GetServerPresenter mGetServerPresenter;

    @InjectPresenter
    CheckSessionPresenter mCheckSessionPresenter;

    @InjectPresenter
    SignOutPresenter mSignOutPresenter;

    @Inject
    NavigatorHolder navigatorHolder;

    @Inject
    Router router;

    @Inject
    SharedPreferences sharedPreferences;

    @Inject
    Bus bus;

    private AlertDialog mErrorDialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        MainApp.getAppComponent().inject(this);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ButterKnife.bind(this);

        setSupportActionBar(mToolbar);

        LinearLayout linearLayout = (LinearLayout) navigationView.getHeaderView(0);
        mLabelName = (TextView) linearLayout.getChildAt(1);
        mLabelEmail = (TextView) linearLayout.getChildAt(2);



        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, mToolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.addDrawerListener(toggle);
        toggle.syncState();

        navigationView.setNavigationItemSelectedListener(this);

        navigatorHolder.setNavigator(navigator);
        mGetServerPresenter.getServer();

        Timber.d("CURRENT USER: " + getIntent().getParcelableExtra("session"));
    }

    @Override
    public void onResume() {
        super.onResume();
        bus.register(this);
    }

    @Override
    public void onPause() {
        bus.unregister(this);
        super.onPause();
    }

    @Override
    protected void onDestroy() {
        navigatorHolder.removeNavigator();
        super.onDestroy();
    }

    @Override
    public void onBackPressed() {
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        // Handle navigation view item clicks here.
        User session = getIntent().getParcelableExtra("session");
        Server server = getIntent().getParcelableExtra("server");
        Bundle args = new Bundle();
        args.putParcelable("session", session);
        args.putParcelable("server", server);
        switch (item.getItemId()) {
            case R.id.nav_map:
                router.newRootScreen(Screens.MAP_SCREEN, args);
                break;
            case R.id.nav_devices:
                router.newRootScreen(Screens.DEVICES_SCREEN, args);
                break;
            case R.id.nav_users:
                router.newRootScreen(Screens.USERS_SCREEN, args);
                break;
            case R.id.nav_server:
                router.navigateTo(Screens.SERVER_SCREEN, args);
                break;
            case R.id.nav_account:
                args.putLong("userId", session.getId());
                router.navigateTo(Screens.USER_SCREEN, args);
                break;
            case R.id.nav_about:
                router.navigateTo(Screens.ABOUT_SCREEN);
                break;
            case R.id.nav_sign_out:
                mSignOutPresenter.signOut();
                break;
        }

        drawer.closeDrawer(GravityCompat.START);
        return true;
    }

    private Navigator navigator = new SupportFragmentNavigator(getSupportFragmentManager(),
            R.id.container) {
        @Override
        protected Fragment createFragment(String screenKey, Object data) {
            switch(screenKey) {
                case Screens.SIGN_IN_SCREEN:
                    return SignInFragment.getNewInstance(data);
                case Screens.SIGN_UP_SCREEN:
                    return SignUpFragment.getNewInstance(data);
                case Screens.MAP_SCREEN:
                    return MapFragment.getNewInstance((Bundle) data);
                case Screens.DEVICES_SCREEN:
                    return DevicesFragment.getNewInstance((Bundle) data);
                case Screens.USERS_SCREEN:
                    return UsersFragment.getNewInstance((Bundle) data);
                case Screens.SERVER_SCREEN:
                    return ServerFragment.getNewInstance((Bundle) data);
                case Screens.USER_SCREEN:
                    return UserFragment.getNewInstance((Bundle) data);
                case Screens.DEVICE_SCREEN:
                    return DeviceFragment.getNewInstance((Bundle) data);
                case Screens.PERMISSIONS_SCREEN:
                    return PermissionsFragment.getNewInstance((Bundle) data);
                case Screens.REPORT_SCREEN:
                    return ReportFragment.getNewInstance((Bundle) data);
                case Screens.ABOUT_SCREEN:
                    return AboutFragment.getNewInstance(data);
                default:
                    throw new RuntimeException("Unknown screen key!");
            }
        }

        @Override
        protected void showSystemMessage(String message) {
            Toast.makeText(MainActivity.this, message, Toast.LENGTH_SHORT).show();
        }

        @Override
        protected void exit() {
            finish();
        }
    };

    private void toggleProgressVisibility(final boolean show) {
        mProgress.setVisibility(show ? View.VISIBLE : View.GONE);
        mContainer.setVisibility(show ? View.GONE : View.VISIBLE);
    }

    @Override
    public void startCheckSession() {
        toggleProgressVisibility(true);
    }

    @Override
    public void finishCheckSession() {
        toggleProgressVisibility(false);
    }

    @Override
    public void failedCheckSession(String message) {
        mToolbar.setVisibility(View.GONE);
        router.newRootScreen(Screens.SIGN_IN_SCREEN);
    }

    @Override
    public void startSignOut() {
        toggleProgressVisibility(true);
    }

    @Override
    public void finishSignOut() {
        toggleProgressVisibility(false);
    }

    @Override
    public void startGetServer() {
        toggleProgressVisibility(true);
    }

    @Override
    public void finishGetServer() {
        toggleProgressVisibility(false);
    }

    @Override
    public void failedGetServer(String message) {
        mErrorDialog = new AlertDialog.Builder(this)
                .setTitle(R.string.app_name)
                .setMessage(message)
                .setOnCancelListener(dialog -> mGetServerPresenter.onErrorCancel())
                .show();

        router.newRootScreen(Screens.SIGN_IN_SCREEN);
    }

    @Override
    public void hideError() {
        if (mErrorDialog != null && mErrorDialog.isShowing()) {
            mErrorDialog.cancel();
        }
    }


    @Override
    public void successGetServer(Server server) {
        getIntent().putExtra("server", server);
        BingMapTileSource.setBingKey(server.getBingKey());
        mCheckSessionPresenter.check();
    }

    @Override
    public void successSignOut() {
        sharedPreferences.edit()
                .putLong("userId", -1)
                .putBoolean("save", false)
                .putString("email", "")
                .putString("password", "")
                .apply();
        mToolbar.setVisibility(View.GONE);
        router.newRootScreen(Screens.SIGN_IN_SCREEN);
    }

    @Override
    public void successCheckSession(User user) {
        Timber.d("successCheckSession");
        successSignIn(user);
    }


    @Override
    public void successSignIn(User user) {
        getIntent().putExtra("session", user);

        mToolbar.setVisibility(View.VISIBLE);
        mLabelName.setText(user.getName());
        mLabelEmail.setText(user.getEmail());

        navigationView.getMenu().findItem(R.id.nav_users).setVisible(user.getAdmin());
        navigationView.getMenu().findItem(R.id.nav_server).setVisible(user.getAdmin());

        router.newRootScreen(Screens.MAP_SCREEN);
    }
}

File: erlymon-monitor-app/src/main/kotlin/org/erlymon/monitor/view/MainActivity.kt
/*
 * Copyright (c) 2016, Sergey Penkovsky <sergey.penkovsky@gmail.com>
 *
 * This file is part of Erlymon Monitor.
 *
 * Erlymon Monitor is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Erlymon Monitor is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Erlymon Monitor.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.erlymon.monitor.view

import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.NavigationView
import android.support.v4.view.GravityCompat
import android.support.v4.view.ViewPager
import android.support.v7.app.ActionBarDrawerToggle
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import io.realm.RealmResults
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.app_bar_main.*
import kotlinx.android.synthetic.main.content_main.*
import org.erlymon.core.model.data.*
import org.erlymon.core.presenter.MainPresenter
import org.erlymon.core.presenter.MainPresenterImpl
import org.erlymon.core.view.MainView
import org.erlymon.monitor.R
import org.erlymon.monitor.view.adapter.CustomFragmentPagerAdapter
import org.erlymon.monitor.view.adapter.DevicesAdapter
import org.erlymon.monitor.view.adapter.UsersAdapter
import org.erlymon.monitor.view.fragment.ConfirmDialogFragment
import org.erlymon.monitor.view.fragment.SendCommandDialogFragment
import org.osmdroid.util.GeoPoint

import org.slf4j.LoggerFactory

class MainActivity : BaseActivity<MainPresenter>(),
        MainView,
        NavigationView.OnNavigationItemSelectedListener,
        DevicesFragment.OnActionDeviceListener,
        UsersFragment.OnActionUserListener,
        ConfirmDialogFragment.ConfirmDialogListener,
        SendCommandDialogFragment.SendCommandDialogListener {

    private var pagerAdapter: CustomFragmentPagerAdapter? = null

    private var mAccountNameView: TextView? = null
    private var mAccountEmailView: TextView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        setSupportActionBar(toolbar)

        presenter = MainPresenterImpl(this, this)

        val linearLayout = nav_view.getHeaderView(0) as LinearLayout
        mAccountNameView = linearLayout.getChildAt(1) as TextView
        mAccountEmailView = linearLayout.getChildAt(2) as TextView

        val session = intent.getParcelableExtra<User>("session")
        mAccountNameView?.text = session?.name
        mAccountEmailView?.text = session?.email

        val toggle = ActionBarDrawerToggle(
                this, drawer_layout, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        drawer_layout.addDrawerListener(toggle)
        toggle.syncState()

        nav_view.setNavigationItemSelectedListener(this)

        (nav_view.menu.findItem(R.id.nav_users) as MenuItem).isVisible = session?.admin!!
        (nav_view.menu.findItem(R.id.nav_server) as MenuItem).isVisible = session?.admin!!


        pagerAdapter = CustomFragmentPagerAdapter(supportFragmentManager)
        pagerAdapter?.addPage(MapFragment())
        pagerAdapter?.addPage(DevicesFragment())
        pagerAdapter?.addPage(UsersFragment())
        view_pager.setAdapter(pagerAdapter)

        view_pager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {

            override fun onPageSelected(position: Int) {
                logger.debug("onPageSelected, position = " + position)
                when (position) {
                    0 -> {
                        fab.visibility = View.GONE
                        supportActionBar?.setTitle(R.string.mapTitle)
                    }
                    1 -> {
                        fab.visibility = View.VISIBLE
                        supportActionBar?.setTitle(R.string.settingsDevices)
                    }
                    2 -> {
                        fab.visibility = View.VISIBLE
                        supportActionBar?.setTitle(R.string.settingsUsers)
                    }
                }
            }

            override fun onPageScrolled(position: Int, positionOffset: Float,
                                        positionOffsetPixels: Int) {
            }

            override fun onPageScrollStateChanged(state: Int) {
            }
        })

        fab.setOnClickListener {
            when (view_pager.currentItem) {
                1 -> {
                    logger.debug("Start DeviceActivity")
                    val intent = Intent(this@MainActivity, DeviceActivity::class.java)
                            .putExtra("session", intent.getParcelableExtra<User>("session"))
                    startActivity(intent)
                }
                2 -> {
                    logger.debug("Start UserActivity")
                    val intent = Intent(this@MainActivity, UserActivity::class.java)
                            .putExtra("session", intent.getParcelableExtra<User>("session"))
                    startActivity(intent)
                }
            }
        }
    }

    override fun onBackPressed() {
        if (drawer_layout.isDrawerOpen(GravityCompat.START)) {
            drawer_layout.closeDrawer(GravityCompat.START)
        } else {
            //super.onBackPressed();
            if (backPressed + 2000 > System.currentTimeMillis()) {
                presenter?.onDeleteSessionButtonClick()
            } else {
                Toast.makeText(baseContext, getString(R.string.sharedBackPressed), Toast.LENGTH_SHORT).show()
                backPressed = System.currentTimeMillis()
            }
        }
    }

    override fun showPosition(position: Position) {
        try {
            (pagerAdapter?.getItem(0) as MapFragment).animateTo(GeoPoint(position.latitude, position.longitude), 15)
            view_pager.setCurrentItem(0)
            nav_view.setCheckedItem(R.id.nav_map)
        } catch (e: Exception) {
            logger.warn(Log.getStackTraceString(e))
        }
    }

    override fun showCompleted() {
        finish()
    }

    override fun showRemoveDeviceCompleted() {
        intent.putExtra("deviceId", 0)
    }

    override fun showRemoveUserCompleted() {
        intent.putExtra("userId", 0)
    }

    override fun getUserId(): Long {
        return intent.getLongExtra("userId", 0)
    }

    override fun getDeviceId(): Long {
        return intent.getLongExtra("deviceId", 0)
    }

    override fun getPositionId(): Long {
        return intent.getLongExtra("positionId", 0)
    }

    override fun getCommand(): Command {
        val command = intent.getParcelableExtra<Command>("command")
        command.deviceId = deviceId
        return command
    }

    override fun showError(error: String) {
        makeToast(toolbar, error)
    }

    @SuppressWarnings("StatementWithEmptyBody")
    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        // Handle navigation view item clicks here.
        when (item.itemId) {
            R.id.nav_map -> {
                view_pager.setCurrentItem(0)
                val center = calculateMapCenter()
                (pagerAdapter?.getItem(0) as MapFragment).animateTo(center.first, center.second)
            }
            R.id.nav_devices -> {
                view_pager.setCurrentItem(1)
            }
            R.id.nav_users -> {
                view_pager.setCurrentItem(2)
            }
            R.id.nav_server -> {
                val intent = Intent(this@MainActivity, ServerActivity::class.java)
                        .putExtra("session", intent.getParcelableExtra<User>("session"))
                        .putExtra("server", intent.getParcelableExtra<Server>("server"))
                startActivityForResult(intent, REQUEST_CODE_UPDATE_SERVER)
            }
            R.id.nav_account -> {
                val intent = Intent(this@MainActivity, UserActivity::class.java)
                        .putExtra("session", intent.getParcelableExtra<User>("session"))
                        .putExtra("user", intent.getParcelableExtra<User>("session"))
                startActivityForResult(intent, REQUEST_CODE_UPDATE_ACCOUNT)
            }
            R.id.nav_about -> {
                startActivity(Intent(this@MainActivity, AboutActivity::class.java))
            }
            R.id.nav_sign_out -> {
                presenter?.onDeleteSessionButtonClick()
            }
        }

        drawer_layout.closeDrawer(GravityCompat.START)
        return true
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when (requestCode) {
            REQUEST_CODE_UPDATE_SERVER ->
                if (resultCode == RESULT_OK) {
                    val server = data?.getParcelableExtra<Server>("server")
                    intent.putExtra("server", server)
                }
            REQUEST_CODE_UPDATE_ACCOUNT ->
                if (resultCode == RESULT_OK) {
                    val user = data?.getParcelableExtra<User>("user")
                    intent.putExtra("session", user)
                    mAccountNameView?.text = user?.name
                    mAccountEmailView?.text = user?.email
                }
            REQUEST_CODE_CREATE_OR_UPDATE_DEVICE ->
                if (resultCode == RESULT_OK) {
                }
        }
    }

    override fun onEditDevice(device: Device) {
        val intent = Intent(this@MainActivity, DeviceActivity::class.java)
                .putExtra("session", intent.getParcelableExtra<User>("session"))
                .putExtra("device", device)
        startActivity(intent)
    }

    override fun onRemoveDevice(device: Device) {
        intent.putExtra("deviceId", device.id)
        val dialogFragment = ConfirmDialogFragment.newInstance(R.string.deviceTitle, R.string.sharedRemoveConfirm)
        dialogFragment.show(supportFragmentManager, "remove_item_dialog")
    }

    override fun onLoadPositions(device: Device) {
        val intent = Intent(this@MainActivity, PositionsActivity::class.java)
                .putExtra("session", intent.getParcelableExtra<User>("session"))
                .putExtra("device", device)
        startActivity(intent)
    }

    override fun onShowOnMap(device: Device) {
        intent.putExtra("positionId", device.positionId)
        presenter?.onGetPostionByCache()
    }

    override fun onSendCommand(device: Device) {
        intent.putExtra("deviceId", device.id)
        val dialogFragment = SendCommandDialogFragment.newInstance(device.id)
        dialogFragment.show(supportFragmentManager, "send_command_dialog")
    }

    override fun onEditUser(user: User) {
        val intent = Intent(this@MainActivity, UserActivity::class.java)
                .putExtra("session", intent.getParcelableExtra<User>("session"))
                .putExtra("user", user)
        startActivity(intent)
    }

    override fun onRemoveUser(user: User) {
        intent.putExtra("userId", user.id)
        val dialogFragment = ConfirmDialogFragment.newInstance(R.string.deviceTitle, R.string.sharedRemoveConfirm)
        dialogFragment.show(supportFragmentManager, "remove_item_dialog")
    }

    override fun onPermissionForUser(user: User) {
        val intent = Intent(this@MainActivity, DevicesActivity::class.java)
                .putExtra("session", intent.getParcelableExtra<User>("session"))
                .putExtra("user", user)
        startActivity(intent)
    }

    override fun onPositiveClick(dialog: DialogInterface, which: Int) {
        if (deviceId > 0) {
            presenter?.onDeleteDeviceButtonClick()
        }
        if (userId > 0) {
            presenter?.onDeleteUserButtonClick()
        }
    }

    override fun onSendCommand(command: Command?) {
        intent.putExtra("command", command)
        presenter?.onSendCommandButtonClick()
    }

    private fun calculateMapCenter() :Pair<GeoPoint, Int> {
        val user = intent.getParcelableExtra<User>("session")
        if (user.latitude === 0.0 && user.longitude === 0.0 && user.zoom === 0) {
            val server = intent.getParcelableExtra<Server>("server")
            if (server.latitude === 0.0 && server.longitude === 0.0 && server.zoom === 0) {
                return Pair(GeoPoint(server.latitude, server.longitude), server.zoom)
            }
        } else {
            return Pair(GeoPoint(user.latitude, user.longitude), user.zoom)
        }
        return Pair(GeoPoint(0, 0), 0)
    }

    companion object {
        private val logger = LoggerFactory.getLogger(MainActivity::class.java)
        private var backPressed: Long = 0

        val REQUEST_CODE_UPDATE_SERVER = 1
        val REQUEST_CODE_UPDATE_ACCOUNT = 2
        val REQUEST_CODE_CREATE_OR_UPDATE_DEVICE = 3
    }
}
